### PR TITLE
Fix case where null start index or length would throw invalid parameter, instead of returning null rows

### PR DESCRIPTION
**Pull request by abellina**
Fixes https://github.com/NVIDIA/spark-rapids/issues/13164

This fixes implementation and test issues with `array_slice` where we were not taking into account null columns or scalars in the `list`, `start` arguments to `array_slice`. The test that failed CI intermittently would do so when it expected the `array_slice` function to fail when `length` was negative, but it didn't, instead returning null rows. The reason for this was because either the rows in `list` or `start` could be null, and in these scenario spark returns null for the result of slice, and doesn't throw.

The fix is to check for these rows, and we use a function to "merge nulls" onto the `length` column, if length is not a scalar. We do this because there could be nulls at precisely the rows where length is negative, and therefore no exception should be thrown.

The patch adds tests to go through more edge cases than before.

### CODE REVIEW

File: integration_tests/src/main/python/array_test.py [hunk 1]
```diff
@@ -388,19 +391,67 @@ def test_array_slice_with_negative_length(data_gen, valid_start, negative_length
...
     assert_gpu_and_cpu_are_equal_collect(
-        lambda spark: three_col_df(spark, array_all_null_gen, array_no_zero_index_gen, negative_length_gen, length=5).selectExpr(
-            f"slice(a, {valid_start}, {negative_length})"))
+        lambda spark: three_col_df(spark,
+                array_all_null_gen, array_no_zero_no_null_index_gen, negative_length_gen,
+                length=5)
+            .selectExpr(f"slice(a, {valid_start}, {negative_length})"))
+
+    # A null column for start, results in nulls for GPU/CPU (no exceptions)
+    assert_gpu_and_cpu_are_equal_collect(
```
Comment (revans2) [inline integration_tests/src/main/python/array_test.py:402]: nit: could we split this up into separate tests? For me having a single test that covers 6 different situations is both confusing and can mask issues if one test fails and blocks the other tests from running.

File: integration_tests/src/main/python/asserts.py [hunk 2]
```diff
@@ -678,34 +678,79 @@ def do_it_all(spark):
...
     if isinstance(error_message, re.Pattern):
         assert error_message.search(actual_error), f"Expected error '{error_message}' to match '{actual_error}'"
     elif isinstance(error_message, str):
         assert error_message in actual_error, f"Expected error '{error_message}' did not appear in '{actual_error}'"
     else:
         assert False, f"expected str or Pattern found {type(error_message)} {error_message}"
 
-def assert_gpu_and_cpu_error(df_fun, conf, error_message):
+
+def assert_spark_exception(func, error_message, expect_exception=None):
```
Comment (revans2) [inline integration_tests/src/main/python/asserts.py]: This is not an assert. This is more of a collect_data_or_exception to me.

File: integration_tests/src/main/python/asserts.py [hunk 3]
```diff
@@ -678,34 +678,79 @@ def do_it_all(spark):
...
+    return (False, result)
+
+
+def assert_gpu_and_cpu_error(df_fun, conf, error_message, only_if_cpu_fails=False):
     """
     Assert that GPU and CPU execution results in a specific Java exception thrown
     :param df_fun: a function to be verified
     :param conf: Spark config
     :param error_message: a string such as the one produce by java.lang.Exception.toString
+    :param only_if_cpu_fails: if true, we only expect the GPU job to fail if the CPU job fails,
```
Comment (revans2) [inline integration_tests/src/main/python/asserts.py:746]: nit: for readability I would rather have this boolean not be here, but instead have a new method that would be something more like. `assert_gpu_and_cpu_same_data_or_error`, but I don't really like that name too much either.

**Comment by abellina** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 4]
```diff
@@ -745,6 +745,26 @@ object GpuDivModLike {
       }
     }
   }
+
+  /**
+   * Return a null row at "i" if either "dataCol[i]", "nullFlagCol1[i]" or "nullFlagCol2[i]"
+   * are null, otherwise return the row from the "dataCol".
+   */
+  def mergeNulls(
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala]: nit: for me I think that the following would be slightly cheaper to compute. The probability that dataCol is larger than a boolean column is fairly high.  Also might be interesting to add in a check if any of nullFlag is true before doing the if/else

```
val nullFlag = withResource(nullFlagCol1.isNull) { nullFlag1 =>
  withResource(nullFlagCol2.isNull) { nullFlag2 =>
    nullFlag1.or(nullFlag2)
  }
}
withResource(nullFlag) { _ =>
  val needsToChange = withResource(nullFlag.any) { tmp =>
     tmp.isValid && tmp.getBoolean
  }
  if (needsToChange) {
    withResource(Scalar.fromNull(dataCol.getType)) { nullScalar =>
      nullFlag.ifElse(nullScalar, dataCol)
    }
  } else {
    dataCol.incRefCount()
  }
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 5]
```diff
@@ -188,30 +188,36 @@ case class GpuSlice(x: Expression, start: Expression, length: Expression)
...
-      val length = lengthCol.getBase
       if (start == 0) {
         throw RapidsErrorUtils.unexpectedValueForStartInFunctionError(prettyName)
       }
-      withResource(length.min()) { minLen =>
-        if (minLen.isValid && minLen.getInt < 0) {
-          throw RapidsErrorUtils.unexpectedValueForLengthInFunctionError(prettyName,
-            minLen.getInt)
+      // keep null as original, or null it out if `list[i]`  is null
+      withResource(GpuDivModLike.mergeNulls(lengthCol.getBase, list)) { length =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala]: nit: would it be better to move this to a different utility class? a collectionOperation is not really a DivModLike operation.

**Review by revans2 — APPROVED** [review]
Everything I have really are nits

**Comment by abellina** [comment]
Thanks @revans2 let me fix

File: integration_tests/src/main/python/asserts.py (hunk 2 above)
Comment (abellina) [reply integration_tests/src/main/python/asserts.py]: well it does assert, and calls the underlying assertion logic that we used to call. But let me rephrase the name.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
whoops broke CI, will fix.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@revans2 @firestarman this is ready for review.

**Review by firestarman — APPROVED** [review]
LGTM

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 6]
```diff
@@ -188,30 +188,36 @@ case class GpuSlice(x: Expression, start: Expression, length: Expression)
...
       val start = startS.getValue.asInstanceOf[Int]
-      val length = lengthCol.getBase
       if (start == 0) {
         throw RapidsErrorUtils.unexpectedValueForStartInFunctionError(prettyName)
       }
-      withResource(length.min()) { minLen =>
-        if (minLen.isValid && minLen.getInt < 0) {
-          throw RapidsErrorUtils.unexpectedValueForLengthInFunctionError(prettyName,
-            minLen.getInt)
+      // keep null as original, or null it out if `list[i]`  is null
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala:202]: I think the comment was ok, and length is implied (as original)

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 7]
```diff
@@ -231,24 +240,29 @@ case class GpuSlice(x: Expression, start: Expression, length: Expression)
...
         if (start.contains(zero)) {
           throw RapidsErrorUtils.unexpectedValueForStartInFunctionError(prettyName)
         }
       }
-      withResource(length.min()) { minLen =>
-        if (minLen.isValid && minLen.getInt < 0) {
-          throw RapidsErrorUtils.unexpectedValueForLengthInFunctionError(prettyName,
-            minLen.getInt)
+      // keep null as original, or null it out if `list[i]` or `start[i]` is null
+      withResource(NullUtilities.mergeNulls(lengthCol.getBase, list, start)) { length =>
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala:257]: lengthCol is implied, comment was ok.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/nullExpressions.scala [hunk 8]
```diff
@@ -313,3 +313,54 @@ case class GpuNaNvl(left: Expression, right: Expression) extends GpuBinaryExpres
 //  override def inputTypes: Seq[AbstractDataType] =
 //    Seq(TypeCollection(DoubleType, FloatType), TypeCollection(DoubleType, FloatType))
 }
+
+object NullUtilities {
+  /**
+   * Return a null row at "i" if either "dataCol[i]" or "nullFlagCol[i]" are null, otherwise
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/nullExpressions.scala:319]: the docstring could be rewarded a little, but it is not incorrect.

File: integration_tests/src/main/python/asserts.py [hunk 9]
```diff
@@ -678,34 +678,91 @@ def do_it_all(spark):
...
+def collect_data_or_exception(func, error_message, expect_exception=None):
     """
-    Assert that GPU and CPU execution results in a specific Java exception thrown
+    Assert that a specific Java exception is thrown
+    :param func: a function to be verified
+    :param error_message: a string such as the one produce by java.lang.Exception.toString
+    :param expect_exception:
+            - if none, we check the error if and only if it was raised
+            - if true, we will assert False if func() doesn't raise an exception,
+            - if false, we fail if we did not raise an exception
```
[the remark this answers is not in the document]
Comment (abellina) [reply integration_tests/src/main/python/asserts.py:707]: Missed this comment, and this one is correct. I'll open a PR to fix.
